Adds `spp model get` command Closes #6105

mkm17 opened the pull request:
Adds `spp model get` command Closes #6105

milanholemans commented:
Thank you, well try to review it ASAP!

mkm17 commented:
I have converted it to draft, as there are some conflicts to be resolved.

Adam-it commented:
@mkm17 I added the `hacktoberfest-accepted` label to this PR which means that this PR will count as done for the Hacktoberfest event. So if you participate in this event it will get you unblocked and it allows us to merge this PR later when we catch up 👍
Thanks for your support and awesome contribution 👏 You Rock 🤩

milanholemans commented on docs/docs/cmd/spp/model/model-get.mdx:
```diff
@@ -0,0 +1,308 @@
+: The URL of the content center site.
+
+`-i, --id [id]`
+: The unique ID of the model to retrieve. Specify id or title but not both.
+
+`-t, --title [title]`
+: The display name (case-sensitive) of the model to retrieve. Specify id or title but not both.
```
```suggestion
: The unique ID of the model to retrieve. Specify `id` or `title` but not both.

`-t, --title [title]`
: The display name (case-sensitive) of the model to retrieve. Specify `id` or `title` but not both.
```

milanholemans commented on docs/docs/cmd/spp/model/model-get.mdx:
````diff
@@ -0,0 +1,308 @@
+m365 spp model get --siteUrl "https://contoso.sharepoint.com/sites/ContentCenter" --title "climicrosoft365Model.classifier"
+```
+
+## Response
````
Let's add a subtitle `Standard response`

milanholemans commented on docs/docs/cmd/spp/model/model-get.mdx:
````diff
@@ -0,0 +1,308 @@
+
+  ```csv
+    AIBuilderHybridModelType,AzureCognitivePrebuiltModelName,BaseContentTypeName,ConfidenceScore,ContentTypeGroup,ContentTypeId,ContentTypeName,Created,CreatedBy,DriveId,Explanations,ID,LastTrained,ListID,ModelSettings,ModelType,Modified,ModifiedBy,ObjectId,PublicationType,Schemas,SourceSiteUrl,SourceUrl,SourceWebServerRelativeUrl,UniqueId
+    ,,,"{""trainingStatus"":{""kind"":""original"",""ClassifierStatus"":{""TrainingStatus"":""success"",""TimeStamp"":1604429035541},""ExtractorsStatus"":[{""TimeStamp"":1595871764667,""ExtractorName"":""Adressee"",""TrainingStatus"":""success""},{""TimeStamp"":1595871896822,""ExtractorName"":""Discount Percentage"",""TrainingStatus"":""success""},{""TimeStamp"":1595871961934,""ExtractorName"":"" ...
````
Let's remove the indent.

milanholemans commented on docs/docs/cmd/spp/model/model-get.mdx:
```diff
@@ -0,0 +1,308 @@
+
+### `listPublications` response
+
+When we make use of the option `listPublications` the response will differ. 
```
```suggestion
When we make use of the option `listPublications`, the response will differ. 
```

milanholemans commented on docs/docs/cmd/spp/model/model-get.mdx:
````diff
@@ -0,0 +1,308 @@
+
+  ```csv
+    AIBuilderHybridModelType,AzureCognitivePrebuiltModelName,BaseContentTypeName,ConfidenceScore,ContentTypeGroup,ContentTypeId,ContentTypeName,Created,CreatedBy,DriveId,Explanations,ID,LastTrained,ListID,ModelSettings,ModelType,Modified,ModifiedBy,ObjectId,PublicationType,Schemas,SourceSiteUrl,SourceUrl,SourceWebServerRelativeUrl,UniqueId
+    ,,,"{""trainingStatus"":{""kind"":""original"",""ClassifierStatus"":{""TrainingStatus"":""success"",""TimeStamp"":1604429035541},""ExtractorsStatus"":[{""TimeStamp"":1595871764667,""ExtractorName"":""Adressee"",""TrainingStatus"":""success""},{""TimeStamp"":1595871896822,""ExtractorName"":""Discount Percentage"",""TrainingStatus"":""success""},{""TimeStamp"":1595871961934,""ExtractorName"":"" ...
````
Let's remove the indent.

milanholemans reviewed (changes requested):
Nice start @mkm17, let's change a few things before we proceed.

mkm17 commented on src/m365/spp/commands/model/model-get.ts:
```diff
@@ -0,0 +1,130 @@
+      let requestUrl = `${siteUrl}/_api/machinelearning/models/`;
+
+      if (args.options.title) {
+        requestUrl += `getbytitle('${formatting.encodeQueryParameter(args.options.title!)}')`;
```
I have added a condition to check if the user includes the '.classifier' suffix in the title. If it's missing, the suffix will be automatically added at the end of the request.

mkm17 commented:
@milanholemans thank you again for the review. I believe I have addressed all the suggested changes. I have one question regarding the tests: I updated it to `assert.strictEqual(JSON.stringify(loggerLogSpy.lastCall.args[0]), JSON.stringify(modelResult));` instead of the previous approach. Is that ok?

milanholemans commented:
> @milanholemans thank you again for the review. I believe I have addressed all the suggested changes. I have one question regarding the tests: I updated it to `assert.strictEqual(JSON.stringify(loggerLogSpy.lastCall.args[0]), JSON.stringify(modelResult));` instead of the previous approach. Is that ok?

Is there a reason why we stringify everything instead of using `assert.deepStrictEqual`?

mkm17 commented:
@milanholemans I saw this method somewhere in the code, but definitely it is a much better idea. Thank you.

milanholemans commented on docs/docs/cmd/spp/model/model-get.mdx:
````diff
@@ -0,0 +1,306 @@
+m365 spp model get --siteUrl "https://contoso.sharepoint.com/sites/ContentCenter" --title "climicrosoft365Model"
+```
+
+### Standard response
````
This should be a subtitle of the `Response` header.

milanholemans commented on docs/docs/cmd/spp/model/model-get.mdx:
```diff
@@ -0,0 +1,306 @@
+    "ModifiedBy": "i:0#.w|sharepoint\\system",
+    "ObjectId": "01XHOXSCXFBUHLC2OA7FDJB52PXCWAAE3S",
+    "PublicationType": 0,
+    "Schemas": {"Extractors":{"Adressee":{"concepts":{"0a18ef8b-c0cd-435d-8a45-37738fef4f6f":{"name":"Adressee"}},"relationships":[]},"Discount Percentage":{"concepts":{"48c09840-d2da-4416-a02d-68113da24d18":{"name":"Discount Percentage"}},"relationships":[]},"Date of Change":{"concepts":{"f39b4e2a-e945-4a2e-99a0-561334037a66":{"name":"Date of Change"}},"relationships":[]},"Date Completed":{"conc ...
```
For all properties that contain a JSON object, let's format them the usual way instead of putting everything on one line.

milanholemans commented on docs/docs/cmd/spp/model/model-get.mdx:
```diff
@@ -0,0 +1,306 @@
+
+When we make use of the option `withPublications`, the response will differ. 
+
+## Response
```
Let's remove this header.

milanholemans commented on src/utils/spp.ts:
```diff
@@ -1,10 +1,20 @@
+  UniqueId: string;
+  Publications?: any[];
+  ModelSettings?: string;
+  'odata.null'?: boolean;
```
This property is not really part of an `SppModel`. I suggest we remove it from here. In the command when we check for `odata.null`, let's just cast the result to `any` in the if-statement.

`if ((result as any)['odata.null'] === true) { .... }`

milanholemans reviewed (changes requested):
It's looking good, we're almost there.

mkm17 commented:
@milanholemans  Thank you for the review. The changes have been applied to the new version.

milanholemans commented on docs/docs/cmd/spp/model/model-get.mdx:
````diff
@@ -0,0 +1,994 @@
+<Tabs>
+  <TabItem value="JSON">
+
+  ```json
````
This response is looking gigantic right now. Let's try to trim it down a bit by:
* For arrays, let's show only 1 object in the array.
* Right now, we have multiple fields in the model like: address, discount percentage, date of change, ... Let's show only one of those.

mkm17 replied:
@milanholemans  ok, it is done now, the response should be shorter :)

milanholemans reviewed (approved):
Apart from one tiny enhancement, it looks good to me.

milanholemans commented:
Merged manually, thanks!
